snuba/query/mql/parser.py

```diff
@@ -0,0 +1,636 @@
+    args["offset"] = resolved_args["offset"]
+    args["granularity"] = resolved_args["granularity"]
+    args["totals"] = resolved_args["totals"]
+
+    query = LogicalQuery(**args)
+    return query
+
+
+def extract_selected_columns(
+    parsed: Mapping[str, Any],
+    resolved_args: Mapping[str, Any],
+    mql_context: Mapping[str, Any],
+) -> list[SelectedExpression]:
+    selected_columns = []
+    if "aggregate" in parsed:
+        selected_columns.append(parsed["aggregate"])
+    if "groupby" in resolved_args:
+        columns = resolved_args["groupby"]
+        for column in columns:
+            assert isinstance(column, Column)
+            selected_columns.append(
+                SelectedExpression(name=column.alias, expression=column)
+            )
+    return selected_columns
+
+
+def extract_args_from_mql_context(
+    parsed: Mapping[str, Any],
+    mql_context: Mapping[str, Any],
+) -> Mapping[str, Any]:
+    """
+    Extracts all metadata from MQL context, creates the appropriate expressions for them,
+    and returns them in a formatted dictionary.
+
+    Example of serialized MQL context:
+        "mql_context": {
+            "entity": "generic_metrics_distributions"
+            "start": "2023-01-02T03:04:05+00:00",
+            "end": "2023-01-16T03:04:05+00:00",
+            "rollup": {
+                    "orderby": {"column_name": "timestamp", "direction": "ASC"},
+                    "granularity": "3600",
+                    "interval": "3600",
+                    "with_totals": "",
+            },
+            "scope": {
+                    "org_ids": ["1"],
+                    "project_ids": ["11"],
+                    "use_case_id": "transactions",
+            },
+            "limit": "",
+            "offset": "0",
+            "indexer_mappings": {
+                "d:transactions/duration@millisecond": "123456", ...
+            }
+        }
+    """
+    resolved_args = {}
+    filters = []
+    groupbys = []
+    if "indexer_mappings" not in mql_context:
+        raise InvalidQueryException("No indexer mappings specified in MQL context.")
+
+    filters.extend(extract_scope(parsed, mql_context))
+    filters.extend(extract_start_end_time(parsed, mql_context))
+    filters.extend(extract_metric_id(parsed, mql_context))
+    filters.extend(extract_resolved_tag_filters(parsed, mql_context))
+
+    groupbys.extend(extract_resolved_gropupby(parsed, mql_context))
+    order_by, granularity, totals = extract_rollup(parsed, mql_context)
+    limit = extract_limit(mql_context)
+    offset = extract_offset(mql_context)
+
+    resolved_args["filters"] = combine_and_conditions(filters)
+    resolved_args["groupby"] = groupbys
+    resolved_args["order_by"] = order_by
+    resolved_args["granularity"] = granularity
+    resolved_args["totals"] = totals
+    resolved_args["limit"] = limit
+    resolved_args["offset"] = offset
+
+    return resolved_args
+
+
+def extract_metric_id(
```

Comment:
There is a baked in assumption here that I think will make it difficult for us to call the indexer from snuba like we planned.

The indexer mappings are only known if you know what metrics are being queried. To do that you need to parse the query. This code is assuming that you are going to know up front what metrics are going to be queried

My recommendation would be to move this further into the query pipeline so the parser can continue to be a parser

Reply:
I'm not sure if I follow completely. Right now as it is, we do know what metrics are being queried up front because the indexer (resolved in the new metrics layer) tells us (snuba). The request that is being sent to snuba will already contain all the resolved indexer mappings needed.

Is the recommendation to create a separate step in the query pipeline which will temporarily mimic calling the indexer, but in actuality, it uses the indexer_mappings such that it will make the indexer migration smoother?